pkg/member/member_control.go

```diff
@@ -265,16 +272,44 @@ func (m *memberControl) UpdateMember(ctx context.Context, cli client.ClusterClos
 	return m.updateMemberPeerAddress(ctx, cli, membersInfo.Header.GetMemberId())
 }
 
-func (m *memberControl) doPromoteMember(ctx context.Context, member *etcdserverpb.Member, cli client.ClusterCloser) error {
-	memPromoteCtx, cancel := context.WithTimeout(ctx, brtypes.DefaultEtcdConnectionTimeout)
+// RemoveMember removes the member from the etcd cluster.
+func (m *memberControl) RemoveMember(ctx context.Context) error {
+	m.logger.Infof("Removing the %s member from cluster", m.podName)
+
+	cli, err := m.clientFactory.NewCluster()
+	if err != nil {
+		return fmt.Errorf("failed to build etcd cluster client : %v", err)
+	}
+	defer cli.Close()
+
+	ctx, cancel := context.WithTimeout(ctx, brtypes.DefaultEtcdConnectionTimeout)
 	defer cancel()
-	_, err := cli.MemberPromote(memPromoteCtx, member.ID) //Member promote call will succeed only if member is in sync with leader, and will error out otherwise
-	if err == nil {                                       //Member successfully promoted
-		m.logger.Infof("Member %v with ID: %v has been promoted", member.GetName(), member.GetID())
-		return nil
-	} else if errors.Is(err, rpctypes.Error(rpctypes.ErrGRPCMemberNotLearner)) { //Member is not a learner
-		m.logger.Info("Member ", member.Name, " : ", member.ID, " already part of etcd cluster")
+
+	memberInfo, err := cli.MemberList(ctx)
+	if err != nil {
+		return fmt.Errorf("error listing members: %v", err)
+	}
+
+	foundMember := findMember(memberInfo.Members, m.podName)
+	if foundMember == nil {
 		return nil
 	}
-	return err
+
+	return miscellaneous.RemoveMemberFromCluster(ctx, cli, foundMember.GetID(), &m.logger)
```

Comment:
My concern here is that since we're using the same context (with a timeout) for two etcd calls, if the first one takes longer to complete for whatever reason, that will leave less time for the second call which may lead to a premature context cancellation

Reply:
yes, I also thought that but that's why I used `DefaultEtcdConnectionTimeout =30sec` to create context.
Reasons:
1. 30secs seems enough for both API call even if one API call takes 20 secs to get complete.
2. If I used different `context.WithTimeout` for each API call then 30sec for each API call seems large to me.
3. It's more difficult to track .... let me explain:
   - you created a `context.WithTimeout` in `func A()` of 10sec. then `func A()` call `func B()`.
   - In, `func B()` you created a `context.WithTimeout` of 30 secs with parent context of `func A()`.
   - now, even if B's `context.WithTimeout` is larger/smaller ...it doesn’t matter as when A's `ctx` will get cancelled B's `ctx` will also get cancelled.
   To me it's more difficult to debug as dev might be thinking that he has given enough time to context in `func B()` but he is likely to forget about `context.WithTimeout` of parent.

Reply:
Here `ctx, cancel := context.WithTimeout(ctx, brtypes.DefaultEtcdConnectionTimeout)` the variable is shadowing the one that is passed into the function. In general this should be avoided.

If you wish to pass the same context to `RemoveMemberFromCluster` then it means that you wish to complete both `MemberList` and `RemoveMemberFromCluster` in the timeout `DefaultEtcdConnectionTimeout`. If that is the intention then the current usage is correct barring the shadowing. If that is not the intent then you should create a new context.WithTimeout and ensure that the previous one is cancelled.

Reply:
Thanks for the example @ishan16696, that makes sense. 
However in this specific case, the parent all the way up tracks to `context.Background()` and it would be beneficial to pass this parent all the way down to all the other functions and have all `context.WithTimeout` calls use this as the parent context

Reply:
> Here ctx, cancel := context.WithTimeout(ctx, brtypes.DefaultEtcdConnectionTimeout) the variable is shadowing the one that is passed into the function. In general this should be avoided.

Ok, I will rename the ctx.

> If you wish to pass the same context to RemoveMemberFromCluster then it means that you wish to complete both MemberList and RemoveMemberFromCluster in the timeout DefaultEtcdConnectionTimeout. If that is the intention then the current usage is correct

yes, that's my intention.

Reply:
> However in this specific case, the parent all the way up tracks to context.Background()

yes, it is.

>  it would be beneficial to pass this parent all the way down to all the other functions and have all context.WithTimeout calls use this as the parent context

why ? can you please elaborate the benefits.

Reply:
Well, passing the parent context and creating child contexts only from that will be one way of ensuring that a case like you described [here](https://github.com/gardener/etcd-backup-restore/pull/509#discussion_r936435358) won't happen


> yes, that's my intention

That's fair enough then. Since it was your intension to use the same context for both calls, and that the timeout is quite large (30s), it's good

Reply:
I'm have now rename the ctx. PTAL